File: python/codeql/log_injection/benign.py
import logging

from flask import Flask, request


app = Flask(__name__)


def clean_for_log(value: str) -> str:
    return value.replace("\r", "\\r").replace("\n", "\\n")


@app.get("/login-safe")
def login_safe():
    logging.warning("login failed for user=%s", clean_for_log(request.args.get("user", "unknown")))
    return "ok"

File: javascript/codeql/log_injection/benign.js
const express = require("express");

const app = express();

function cleanForLog(value) {
  return String(value).replace(/\r/g, "\\r").replace(/\n/g, "\\n");
}

app.get("/login-safe", (req, res) => {
  console.warn("login failed for user=%s", cleanForLog(req.query.user || "unknown"));
  res.send("ok");
});
